### PR TITLE
[BEAM-7969] Report FnAPI counters as deltas in streaming jobs.

### PR DESCRIPTION
**Pull request by Ardagan**
FnAPI counters were not reported in streaming worker.
Streaming Dataflow jobs do not tie counters to workitems,
but report them as deltas on same workitem instead.

This change picks only latest counters per workitem
from FnAPI harness, treats them as updates and reports
to DFE.

[Tested manually at this point]

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by Ardagan** [comment]
R: @lukecwik, @ajamato

**Comment by aaltay** [comment]
cc: @angoenka

**Comment by Ardagan** [comment]
@rohdesamuel

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 1]
```diff
@@ -1867,6 +1873,28 @@ private void sendWorkerUpdatesToDataflowService(
           cumulativeCounters.extractUpdates(false, DataflowCounterUpdateExtractor.INSTANCE));
       counterUpdates.addAll(
           deltaCounters.extractModifiedDeltaUpdates(DataflowCounterUpdateExtractor.INSTANCE));
+      if (hasExperiment(options, "beam_fn_api")) {
+        while (!this.pendingMonitoringInfos.isEmpty()) {
+          final CounterUpdate item = this.pendingMonitoringInfos.poll();
```
Comment (ajamato) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: The SDK harness will give you the full value of the counter on each MonitoringInfo. Don't you need to convert it to a delta, since the last reported value? Or does setting cumulative have DFE solve this internally?

You might want to confirm with @edre.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 2]
```diff
@@ -1867,6 +1873,28 @@ private void sendWorkerUpdatesToDataflowService(
...
+            item.setCumulative(false);
+          } else {
+            // In current world all counters coming from FnAPI are cumulative.
+            // This is a safety check in case new counter type appears in FnAPI.
+            throw new UnsupportedOperationException(
+                "FnApi counters are expected to provide cumulative values."
+                    + " Please, update convertion to delta logic"
+                    + " if non-cumulative counter type is required.");
+          }
+
```
Comment (ajamato) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: This method essentially has 3 inputs correct? pendingMonitoringInfos, deltaCounters and cumulativeCounters.
And one output counterUpdates

Can you write some unit tests which verifies that counterUpdates is populated properly?

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 3]
```diff
@@ -400,9 +400,13 @@ public int getSize() {
   private final StreamingDataflowWorkerOptions options;
   private final boolean windmillServiceEnabled;
   private final long clientId;
+
   private final MetricTrackingWindmillServerStub metricTrackingWindmillServer;
   private final CounterSet pendingDeltaCounters = new CounterSet();
   private final CounterSet pendingCumulativeCounters = new CounterSet();
+  private final java.util.concurrent.ConcurrentLinkedQueue<CounterUpdate> pendingMonitoringInfos =
```
Comment (ajamato) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: can you please add some comments about how pendingMonitoringInfos is concurrently read/written to? i.e. Which threads are running concurrently and using this

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 1 above)
Comment (edre) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: It's ok to just report the full value as the delta, but only as long as we only report counters when finishing the work item.

Comment (Ardagan) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: Non-cumulative means delta.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 3 above)
Comment (Ardagan) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: done.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 1 above)
Comment (angoenka) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: Lets also check for item == null to avoid possible race condition as its called at start, via a timer and at stop.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 4]
```diff
@@ -1867,6 +1873,28 @@ private void sendWorkerUpdatesToDataflowService(
...
+          // This change will treat counter as delta.
+          // This is required because we receive cumulative results from FnAPI harness,
+          // while streaming job is expected to receive delta updates to counters on same
+          // WorkItem.
+          if (item.getCumulative()) {
+            item.setCumulative(false);
+          } else {
+            // In current world all counters coming from FnAPI are cumulative.
+            // This is a safety check in case new counter type appears in FnAPI.
+            throw new UnsupportedOperationException(
```
Comment (angoenka) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: Shall we log the error and move on?

Comment (Ardagan) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: It is better to fail fast in this case. New type of counter appears only when someone modifies SDK, so it comes from beam developer, not user.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 1 above)
Comment (Ardagan) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: We are good here since we first initialize class and metric and only then start multiple threads.

**Comment by Ardagan** [comment]
> This map uses `MetricKey` as the key which can do not contain sdkHarness id and can potentially have name collision across multiple sdkHarness.
> We use multiple SdkHarness in case of python SDK which will have this problem.

Can you elaborate on the problem a bit please? IIUC different SDK harnesses will handle different steps, so metric keys will not overlap anyway.

Also I feel fixing that issue better be done in separate PR.

**Comment by angoenka** [comment]
Thanks for the offline discussion.
It seems reasonable to move with the PR and add the integration tests.

**Comment by aaltay** [comment]
@Ardagan do you know what the test failures are?

**Comment by Ardagan** [comment]
> @Ardagan do you know what the test failures are?

Seem to be some libraries import failure. Does work on my machine. Looking for fix.

**Comment by Ardagan** [comment]
Run Java PreCommit

**Comment by Ardagan** [comment]
Run Python Dataflow ValidatesContainer

**Comment by angoenka** [comment]
The error seems to be for matching the metrics now. Is this a different error?

**Comment by angoenka** [comment]
For missing `google.cloud` module, Beam tests do not install GCP modules. https://github.com/apache/beam/blob/master/sdks/python/setup.py#L131

**Comment by Ardagan** [comment]
Testing this in separate PR not to pollute for reviewers. Current summary.
Seems that I should add this test to ValidatesRunner suite, since ValidatesContainer suite doesn't override worker jar.
Regarding pubsub: Regular tests don't import pubsub, so what we do is import within scope of method. This triggers import only when test is triggered.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 2 above)
Comment (Ardagan) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: Done

File: sdks/python/apache_beam/runners/dataflow/dataflow_exercise_streaming_metrics_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+import apache_beam as beam
+from apache_beam.metrics import Metrics
+from apache_beam.options.pipeline_options import PipelineOptions
+from apache_beam.options.pipeline_options import SetupOptions
+from apache_beam.options.pipeline_options import StandardOptions
+
+SLEEP_TIME_SECS = 1
+
+
+class StreamingUserMetricsDoFn(beam.DoFn):
```
Comment (ajamato) [inline sdks/python/apache_beam/runners/dataflow/dataflow_exercise_streaming_metrics_pipeline.py]: Pleaes add some comments to describe what this test does,

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 4 above)
Comment (ajamato) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: I think you should not fail in this case. What happens if a someone writes a custom SDK to run on Dataflow. It may introduce different metrics, but you should not fail.

An SDK can also be written by someone outside of beam repo entirely, not necessarily written by a beam developer.

https://docs.google.com/document/d/1MtBZYV7NAcfbwyy9Op8STeFNBxtljxgy69FkHMvhTMA/edit
Customization- Custom SDKs+engines can add their own custom metrics and custom monitored state without modifying beam proto files.

File: sdks/python/apache_beam/runners/dataflow/dataflow_exercise_streaming_metrics_pipeline.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+      help=('Input PubSub subscription of the form '
+            '"projects/<PROJECT>/subscriptions/<SUBSCRIPTION>."'))
+  known_args, pipeline_args = parser.parse_known_args(argv)
+
+  pipeline_options = PipelineOptions(pipeline_args)
+  pipeline_options.view_as(SetupOptions).save_main_session = True
+  pipeline_options.view_as(StandardOptions).streaming = True
+  pipeline = beam.Pipeline(options=pipeline_options)
+
+  _ = (pipeline
```
Comment (ajamato) [inline sdks/python/apache_beam/runners/dataflow/dataflow_exercise_streaming_metrics_pipeline.py]: Is there any reason why you did not reuse the existing pipeline integration test and run it in streaming mode? Why did you need to write a new pipeline? Rather than reusing code.

File: sdks/python/apache_beam/runners/dataflow/dataflow_exercise_streaming_metrics_pipeline_test.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+                  'experiment': 'beam_fn_api',
+                  'input_subscription': self.input_sub.name,
+                  'output_topic': self.output_topic.name,
+                 }
+
+    argv = self.test_pipeline.get_full_options_as_args(**extra_opts)
+    return dataflow_exercise_streaming_metrics_pipeline.run(argv)
+
+  @attr('IT', 'ValidatesRunner')
+  def test_streaming_pipeline_returns_expected_user_metrics_fnapi_it(self):
```
Comment (ajamato) [inline sdks/python/apache_beam/runners/dataflow/dataflow_exercise_streaming_metrics_pipeline_test.py]: add comment describing that this test does.

File: sdks/python/apache_beam/runners/dataflow/dataflow_exercise_streaming_metrics_pipeline.py (hunk 6 above)
Comment (Ardagan) [reply sdks/python/apache_beam/runners/dataflow/dataflow_exercise_streaming_metrics_pipeline.py]: Reading from files requires side inputs that is not properly supported in FnApi streaming and causes existing pipeline to fail. Reusing your existing pipeline was the first thing I tried to do.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 4 above)
Comment (Ardagan) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: That was bad way to formulate on my side:
CounterUpdates are generated within Java DF Worker. So if we receive different type of CounterUpdate here, that means that someone modified DF worker.

**Comment by Ardagan** [comment]
Run Java PreCommit

**Comment by Ardagan** [comment]
run java precommit

**Comment by aaltay** [comment]
Thank you all for fixing this!
